xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,51 @@
-
-#if FLT_RADIX == 2
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 7 + 7) / 8)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 8)
-# define MAX_MANTISSA (1 << (DBL_MAX_EXP & 7))
-#elif FLT_RADIX == 16
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 1 + 1) / 2)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 2)
-# define MAX_MANTISSA (1 << ((DBL_MAX_EXP & 1) * 4))
-#else
-# error FLT_RADIX is a value not currently handled (not 2 or 16)
-// # define MAX_MANTISSA_BYTES (sizeof(double) + 1)
-#endif
-
-static int base256ify_double(double &v) {
-    int exp;
-    v = frexp(v, &exp);
-    // v is now in the range [0.5, 1.0)
-    --exp;
-#if FLT_RADIX == 2
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
-#endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
-}
-
-std::string serialise_double(double v)
+string serialise_double(double v)
 {
-    /* First byte:
-     *  bit 7 Negative flag
-     *  bit 4..6 Mantissa length - 1
-     *  bit 0..3 --- 0-13 -> Exponent + 7
-     *            \- 14 -> Exponent given by next byte
-     *             - 15 -> Exponent given by next 2 bytes
-     *
-     * Then optional medium (1 byte) or large exponent (2 bytes, lsb first)
-     *
-     * Then mantissa (0 iff value is 0)
-     */
-
-    bool negative = (v < 0.0);
-
-    if (negative) v = -v;
-
-    int exp = base256ify_double(v);
-
-    string result;
-
-    if (exp <= 6 && exp >= -7) {
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
+	int exp;
+	frexp(v, &exp);
+	--exp;
+	exp >>= 3;
+	if (exp < -32768 || exp > 32767)
+	{
 		throw Xapian::InternalError("Insane exponent in floating point number");
-	    }
-	    result += negative ? char(0x8f) : char(0x0f);
-	    result += char(unsigned(exp + 32768) & 0xff);
-	    result += char(unsigned(exp + 32768) >> 8);
 	}
-    }
-
-    int maxbytes = min(MAX_MANTISSA_BYTES, 8);
-
-    size_t n = result.size();
-    do {
-	unsigned char byte = static_cast<unsigned char>(v);
-	result += char(byte);
-	v -= double(byte);
-	v *= 256.0;
-    } while (v != 0.0 && --maxbytes);
-
-    n = result.size() - n;
-    if (n > 1) {
-	Assert(n <= 8);
-	result[0] = static_cast<unsigned char>(result[0] | ((n - 1) << 4));
-    }
-
-    return result;
+	int check_if_big_endian = 1;
+	if (*(char *) &check_if_big_endian != 1)
+	{
+		// platform is big-endian
+		uint64_t * temp_cast_pointer =
+			reinterpret_cast<uint64_t *>(&v);
+		*temp_cast_pointer = do_bswap(*temp_cast_pointer);
+		v = *reinterpret_cast<double *>(temp_cast_pointer);
+	}
+	size_t double_size = sizeof(double);
+	char result_temp[double_size];
+	memcpy(result_temp, &v, double_size);
+	string result(result_temp, double_size);
+	return result;
```

Comment:
There's no need to `memcpy()` in the little-endian case - just return a string from a cast pointer to the double.